oauthenticator/google.py

```diff
@@ -124,37 +116,78 @@ def _cast_hosted_domain(self, proposal):
         help="""Google Apps hosted domain string, e.g. My College""",
     )
 
-    async def user_is_authorized(self, auth_model):
-        user_email = auth_model["auth_state"][self.user_auth_state_key]['email']
-        user_email_domain = user_email.split('@')[1]
+    # async def user_is_authorized(self, auth_model):
+    #     """
+    #     Checks that the google user has a verified email and is part of
+    #     `hosted_domain` if set.
+
+    #     Authorizes users part of: `allowed_users`, `admin_users`,
+    #     `allowed_google_groups`, or `admin_google_groups`.
+
+    #     Note that this function also updates the auth_model with admin status
+    #     and the user's google groups if either `allowed_google_groups` or
+    #     `admin_google_groups` are configured.
+    #     """
+    async def check_allowed(self, username, auth_model):
+        """
+        Returns True for users allowed to be authorized.
 
-        if not auth_model["auth_state"][self.user_auth_state_key]['verified_email']:
+        Overrides the OAuthenticator.check_allowed implementation to allow users
+        either part of `allowed_users` or `allowed_organizations`, and not just those
+        part of `allowed_users`.
+        """
+        # allow admin users recognized via admin_users or update_auth_model
+        if auth_model["admin"]:
+            return True
+
+        user_info = auth_model["auth_state"][self.user_auth_state_key]
+        user_email = user_info["email"]
+        user_domain = user_email.split("@")[1]
+        user_groups = set(self._google_groups_for_user(user_email, user_domain))
+        # FIXME: consider overriding the `is_admin` and add this logic there
```

Comment:
```suggestion
```

`is_admin` is only used from `Authenticator.get_authenticated_user`, _after_ a decision has been made if the user is allowed with `check_allowed`. This means that if we put this logic in `Authenticator.is_admin`, it would need to be called also by `check_allowed` first separately.

This is not out of the question, but I think we should instead just put logic updating the relevant things to make decisions on into `update_auth_model`. This aligns with various comments and docstrings:

```python
    async def authenticate(self, handler, data=None, **kwargs):
        # ...

        # update the auth_model with info to later authorize the user in
        # check_allowed, such as admin status and group memberships
        return await self.update_auth_model(auth_model)
```

```python
    async def update_auth_model(self, auth_model):
        """
        Updates and returns the `auth_model` dict.

        Should be overridden to collect information required for check_allowed.

        Args: auth_model - the auth model dictionary, containing:
            - `name`: the normalized username
            - `admin`: the admin status (True/False/None), where None means it
                should be unchanged.
            - `auth_state`: the dictionary of of auth state
                returned by :meth:`oauthenticator.OAuthenticator.build_auth_state_dict`

        Called by the :meth:`oauthenticator.OAuthenticator.authenticate`
        """
        return auth_model
```